python/taichi/lang/ops.py

```diff
@@ -107,83 +108,124 @@ def sqr(obj):
 
 
 @unary
-def neg(expr):
-    return Expr(taichi_lang_core.expr_neg(expr.ptr), tb=stack_info())
-
-
-@unary
-def sin(expr):
-    return Expr(taichi_lang_core.expr_sin(expr.ptr), tb=stack_info())
+def neg(a):
+    if is_taichi_expr(a):
+        return Expr(taichi_lang_core.expr_neg(a.ptr), tb=stack_info())
+    else:
+        return -a
 
 
 @unary
-def cos(expr):
-    return Expr(taichi_lang_core.expr_cos(expr.ptr), tb=stack_info())
+def sin(a):
+    if is_taichi_expr(a):
+        return Expr(taichi_lang_core.expr_sin(a.ptr), tb=stack_info())
+    else:
+        return math.sin(a)
 
 
 @unary
-def asin(expr):
-    return Expr(taichi_lang_core.expr_asin(expr.ptr), tb=stack_info())
+def cos(a):
```

Comment:
Thanks for modifying everything here. The changes have too much repetition and are kind of error-prone. 

Instead of repeating yourself 20+ times in a single file, we should probably consider extracting the common parts.  Something like
```python
def unary_operation(taichi_op, python_op):
    ....

@unary
def cos(a):
    return unary_operation(taichi_lang_core.expr_cos, math.cos)
```

The same for binary operators.

Reply:
I used `_unary_operation` to prevent name issue against the evil `from . import *` pattern.
Btw, is that helpful to make a code generator for this situation to anti-error (next time)?

Reply:
I actually think simply use a function or a better decorator will do the job :-)  A code generator is too complex for the case here.

Reply:
Do you have local commits that are not pushed? I don't see the `_unary_operation` function in the latest code.

Reply:
Oh sorry! Didn't push... too much PRs and I forget what is what.

Reply:
Cool, now everything looks much better. Thanks.